solr/solrj/src/java/org/apache/solr/client/solrj/io/Tuple.java

```diff
@@ -77,38 +77,26 @@ public Tuple() {
    * A copy constructor.
    * @param fields map containing keys and values to be copied to this tuple
    */
-  public Tuple(Map<?, ?> fields) {
-    for (Map.Entry<?, ?> entry : fields.entrySet()) {
-      put(entry.getKey(), entry.getValue());
-    }
+  public Tuple(Map<String, ?> fields) {
+    this.fields.putAll(fields);
+    EOF = this.fields.containsKey(StreamParams.EOF);
+    EXCEPTION = this.fields.containsKey(StreamParams.EXCEPTION);
   }
 
-  /**
-   * Constructor that accepts an even number of arguments as key / value pairs.
-   * @param fields a list of key / value pairs, with keys at odd and values at
-   *               even positions.
-   */
-  public Tuple(Object... fields) {
-    if (fields == null) {
-      return;
-    }
-    if ((fields.length % 2) != 0) {
-      throw new RuntimeException("must have a matching number of key-value pairs");
-    }
-    for (int i = 0; i < fields.length; i += 2) {
-      // skip empty entries
-      if (fields[i] == null) {
-        continue;
-      }
-      put(fields[i], fields[i + 1]);
-    }
+  public Tuple(String k1, Object v1) {
+    if (k1 != null) put(k1, v1);
+  }
+
+  public Tuple(String k1, Object v1, String k2, Object v2) {
+    if (k1 != null) put(k1, v1);
+    if (k2 != null) put(k2, v2);
   }
 
   public Object get(Object key) {
```

Comment:
Thanks for explaining about the _"PECS (producer-extends, consumer-super)"_ mnemonic!

So if I understood that right then that also explains why we don't change `Object key` to `String key` here and in [other similar methods](https://github.com/cpoerschke/solr/commit/220be7aa6ae36b8922f148ae72ab9b74f2478381) right?

Reply:
That one's a little more convoluted because the get and remove and containsKey methods on `Map` take an Object not a generic `K`. Historically this was done for backwards compatibility for something like taking the intersection of a map with Long keys and one with Integer keys for example, which ended up being very hard to do otherwise. That said, Strings are strings, so we could probably change all of the Tuple methods to be safer that way.